xarray/core/indexes.py

```diff
@@ -264,7 +262,7 @@ def sel(self, labels: dict[Any, Any]) -> IndexSelResult:
         """
         raise NotImplementedError(f"{self!r} doesn't support label-based selection")
 
-    def join(self: T_Index, other: T_Index, how: JoinOptions = "inner") -> T_Index:
+    def join(self, other: T_Index, how: JoinOptions = "inner") -> Self:
```

Comment:
```suggestion
    def join(self, other: Self, how: JoinOptions = "inner") -> Self:
```
Not sure if this is supposed to support any indexes again.